cdatransform/extract/idc.py

```diff
@@ -101,88 +122,158 @@ def download_blob(self):
             return temp
         else:
             return """Null""" + """ AS """ + k
         return val
 
     def add_entity_fields(self, entity):
-        entity_string = ''
-        keys = list(self.mapping[entity]['Mapping'].keys())
-        for index in range(len(self.mapping[entity]['Mapping'].keys())):
-            entity_string += self.field_line(keys[index], self.mapping[entity]['Mapping'][keys[index]])
-            if index < len(keys)-1:
+        entity_string = ""
+        keys = list(self.mapping[entity]["Mapping"].keys())
+        for index in range(len(self.mapping[entity]["Mapping"].keys())):
+            entity_string += self.field_line(
+                keys[index], self.mapping[entity]["Mapping"][keys[index]], entity
+            )
+            if index < len(keys) - 1:
                 entity_string += """, """
         return entity_string
 
     def build_where_patients(self):
-        where = ''
+        where = ""
         if self.patient_ids is not None:
             where = """WHERE PatientID in ("""
-            where += """','""".join(self.patient_ids)+"""')"""
+            where += """','""".join(self.patient_ids) + """')"""
         return where
 
+    def create_udf_str(self, func_desc):
+        out = (
+            "CREATE TEMP FUNCTION "
+            + func_desc[0].__name__
+            + "(x "
+            + func_desc[1][0]
+            + ") RETURNS "
+        )
+        out = out + func_desc[1][0] + " AS ("
+        if len(func_desc[1]) == 1:
+            out = out + func_desc[0]()
+        else:
+            out = out + func_desc[0](func_desc[1][1:])
+        out = out + "); "
+        return out
+
+    def create_all_udfs(self):
+        functions_added = []
+        all_udfs = ""
+        for entity, Map_and_Trans in self.mapping.items():
+            if Map_and_Trans.get("Transformations", None) is not None:
+                for field, functions in Map_and_Trans.get("Transformations").items():
+                    for function in functions:
+                        if function[0] not in functions_added:
+                            all_udfs = all_udfs + self.create_udf_str(function)
+        return all_udfs
+
     def _query_build(self, **kwargs):
-        query = """SELECT """
-        query += self.add_entity_fields('Patient')
+        query = self.create_all_udfs()
+        query += """ SELECT """
+        query += self.add_entity_fields("Patient")
         query += """, """
         # add File record structure to query
         query += """ARRAY_AGG(STRUCT("""
-        query += self.add_entity_fields('File')
+        query += self.add_entity_fields("File")
         query += """)) as File """
         # add WHERE statement if just looking for specific patients
         query += """FROM `""" + self.source_table + """`"""
         query += self.build_where_patients()
-        query += """ GROUP by id"""
+        query += """ GROUP by id, species, collection_id"""
+        print(query)
```

Comment:
did you mean to keep this print statment

Reply:
I could remove it if necessary. It just helped me figure out what to fix if the query goes wrong for some reason

Reply:
I understand completely i just wanted to make sure it was not a mistake left in.